src/backend/mod.rs

```diff
@@ -61,35 +61,35 @@ pub trait Backend<'a>: Debug {
     type ListOfListings: MediumCodeGenUnit<()>;
     type Appendix: MediumCodeGenUnit<()>;
 
-    type Paragraph: CodeGenUnit<'a, ()>;
-    type Rule: CodeGenUnit<'a, ()>;
-    type Header: CodeGenUnit<'a, Header<'a>>;
-    type BlockQuote: CodeGenUnit<'a, ()>;
-    type CodeBlock: CodeGenUnit<'a, CodeBlock<'a>>;
-    type List: CodeGenUnit<'a, ()>;
-    type Enumerate: CodeGenUnit<'a, Enumerate>;
-    type Item: CodeGenUnit<'a, ()>;
-    type FootnoteDefinition: CodeGenUnit<'a, FootnoteDefinition<'a>>;
-    type UrlWithContent: CodeGenUnit<'a, Url<'a>>;
-    type InterLinkWithContent: CodeGenUnit<'a, InterLink<'a>>;
-    type HtmlBlock: CodeGenUnit<'a, ()>;
-    type Figure: CodeGenUnit<'a, Figure<'a>>;
-
-    type TableFigure: CodeGenUnit<'a, Figure<'a>>;
-    type Table: CodeGenUnit<'a, Table<'a>>;
-    type TableHead: CodeGenUnit<'a, ()>;
-    type TableRow: CodeGenUnit<'a, ()>;
-    type TableCell: CodeGenUnit<'a, ()>;
-
-    type InlineEmphasis: CodeGenUnit<'a, ()>;
-    type InlineStrong: CodeGenUnit<'a, ()>;
-    type InlineStrikethrough: CodeGenUnit<'a, ()>;
-    type InlineCode: CodeGenUnit<'a, ()>;
-    type InlineMath: CodeGenUnit<'a, ()>;
-
-    type Equation: CodeGenUnit<'a, Equation<'a>>;
-    type NumberedEquation: CodeGenUnit<'a, Equation<'a>>;
-    type Graphviz: CodeGenUnit<'a, Graphviz<'a>>;
+    type Paragraph: StatefulCodeGenUnit<'a, Self, ()>;
+    type Rule: StatefulCodeGenUnit<'a, Self, ()>;
+    type Header: StatefulCodeGenUnit<'a, Self, Header<'a>>;
+    type BlockQuote: StatefulCodeGenUnit<'a, Self, ()>;
+    type CodeBlock: StatefulCodeGenUnit<'a, Self, CodeBlock<'a>>;
+    type List: StatefulCodeGenUnit<'a, Self, ()>;
+    type Enumerate: StatefulCodeGenUnit<'a, Self, Enumerate>;
+    type Item: StatefulCodeGenUnit<'a, Self, ()>;
+    type FootnoteDefinition: StatefulCodeGenUnit<'a, Self, FootnoteDefinition<'a>>;
+    type UrlWithContent: StatefulCodeGenUnit<'a, Self, Url<'a>>;
+    type InterLinkWithContent: StatefulCodeGenUnit<'a, Self, InterLink<'a>>;
+    type HtmlBlock: StatefulCodeGenUnit<'a, Self, ()>;
+    type Figure: StatefulCodeGenUnit<'a, Self, Figure<'a>>;
+
+    type TableFigure: StatefulCodeGenUnit<'a, Self, Figure<'a>>;
+    type Table: StatefulCodeGenUnit<'a, Self, Table<'a>>;
+    type TableHead: StatefulCodeGenUnit<'a, Self, ()>;
+    type TableRow: StatefulCodeGenUnit<'a, Self, ()>;
+    type TableCell: StatefulCodeGenUnit<'a, Self, ()>;
+
+    type InlineEmphasis: StatefulCodeGenUnit<'a, Self, ()>;
+    type InlineStrong: StatefulCodeGenUnit<'a, Self, ()>;
+    type InlineStrikethrough: StatefulCodeGenUnit<'a, Self, ()>;
+    type InlineCode: StatefulCodeGenUnit<'a, Self, ()>;
+    type InlineMath: StatefulCodeGenUnit<'a, Self, ()>;
+
+    type Equation: StatefulCodeGenUnit<'a, Self, Equation<'a>>;
+    type NumberedEquation: StatefulCodeGenUnit<'a, Self, Equation<'a>>;
+    type Graphviz: StatefulCodeGenUnit<'a, Self, Graphviz<'a>>;
```

Comment:
What is the basis to decide which code gen level to require for a specific item kind?

Reply:
Direct elements without content / enclosed events are Simple/Medium code gen units. Elements which can contain other events (namely all tags) can intercept the generated output for contained events and thus are `{Stateful,}CodeGenUnits`.

Reply:
That sounds like a useful distinction that likely deserves a comment in the documentation of the trait.

Reply:
Done.